src/v/cloud_storage/manifest.cc

```diff
@@ -0,0 +1,442 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_storage/manifest.h"
+
+#include "bytes/iobuf.h"
+#include "bytes/iobuf_istreambuf.h"
+#include "bytes/iobuf_ostreambuf.h"
+#include "cluster/types.h"
+#include "hashing/xx.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+#include "model/timestamp.h"
+#include "ssx/sformat.h"
+#include "storage/ntp_config.h"
+
+#include <seastar/core/coroutine.hh>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/lexical_cast.hpp>
+#include <rapidjson/document.h>
+#include <rapidjson/istreamwrapper.h>
+#include <rapidjson/ostreamwrapper.h>
+#include <rapidjson/rapidjson.h>
+#include <rapidjson/writer.h>
+
+#include <algorithm>
+#include <array>
+#include <chrono>
+
+namespace cloud_storage {
+
+manifest::manifest()
+  : _ntp()
+  , _rev()
+  , _last_offset(0) {}
+
+manifest::manifest(model::ntp ntp, model::revision_id rev)
+  : _ntp(std::move(ntp))
+  , _rev(rev)
+  , _last_offset(0) {}
+
+// NOTE: the methods that generate remote paths use the xxhash function
+// to randomize the prefix. S3 groups the objects into chunks based on
+// these prefixes. It also applies rate limit to chunks so if all segments
+// and manifests will have the same prefix we will be able to do around
+// 3000-5000 req/sec. AWS doc mentions that having only two prefix
+// characters should be enough for most workloads
+// (https://aws.amazon.com/blogs/aws/amazon-s3-performance-tips-tricks-seattle-hiring-event/)
+// We're using eight because it's free and because AWS S3 is not the only
+// backend and other S3 API implementations might benefit from that.
+
+static remote_manifest_path generate_partition_manifest_path(
```

Comment:
Might be worth exposing signature in a header file like `manifest_utils.h` and writing a simple test for this, largely to ensure it doesn't change.